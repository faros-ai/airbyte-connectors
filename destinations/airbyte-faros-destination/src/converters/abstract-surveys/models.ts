export type Survey = {
  uid: string;
  name: string;
  description: string;
  type: SurveyType;
  status: SurveyStatus;
  startedAt: Date;
  endedAt: Date;
  creator?: SurveyUser;
  stats?: SurveyStats;
  source: string;
};

export type SurveyQuestion = {
  uid: string;
  question?: string;
  description?: string;
  questionCategory: SurveyQuestionCategoryType;
  responseType: SurveyResponseType;
  source: string;
};

export type SurveyQuestionAssociation = {
  survey: Survey;
  question: SurveyQuestion;
  order: number;
};

export type SurveyStats = {
  questionCount: number;
  invitationCount: number;
  responseCount: number;
};

export type SurveyUser = {
  uid: string;
  email: string;
  name: string;
  source: string;
};

export type SurveyTeam = {
  uid: string;
  name: string;
  description?: string;
  source: string;
};

export type SurveyType = {
  category: SurveyCategory;
  detail: string;
};

export enum SurveyCategory {
  ENPS = 'ENPS',
  NPS = 'NPS',
  Satisfaction = 'Satisfaction',
  Custom = 'Custom',
}

export type SurveyStatus = {
  category: SurveyStatusCategory;
  detail: string;
};

export enum SurveyStatusCategory {
  Completed = 'Completed',
  Canceled = 'Canceled',
  Planned = 'Planned',
  InProgress = 'InProgress',
  Custom = 'Custom',
}

export type SurveyQuestionCategoryType = {
  category: SurveyQuestionCategory;
  detail: string;
};

export enum SurveyQuestionCategory {
  AlignmentAndGoals = 'AlignmentAndGoals',
  DeveloperProductivity = 'DeveloperProductivity',
  ENPS = 'ENPS',
  NPS = 'NPS',
  Quality = 'Quality',
  Satisfaction = 'Satisfaction',
  SpeedAndAgility = 'SpeedAndAgility',
  CodingAssistants = 'CodingAssistants',
  Custom = 'Custom',
}

export type SurveyResponseType = {
  category: SurveyResponseCategory;
  detail: string;
};

export enum SurveyResponseCategory {
  Binary = 'Binary',
  LikertScale = 'LikertScale',
  MultipleChoice = 'MultipleChoice',
  NumericEntry = 'NumericEntry',
  OpenEnded = 'OpenEnded',
  RankOrder = 'RankOrder',
  Rating = 'Rating',
  Custom = 'Custom',
}

export type QuestionCategoryMapping = Record<string, string>;
