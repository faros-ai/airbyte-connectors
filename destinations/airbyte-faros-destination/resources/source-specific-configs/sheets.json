{
  "title": "Sheets",
  "description": "Configuration options that apply to records generated by the Sheets Source.",
  "type": "object",
  "oneOf": [
    {
      "title": "Surveys",
      "type": "object",
      "properties": {
        "application_mapping": {
          "type": "string",
          "title": "Application Mapping",
          "description": "JSON map from survey question categories to Faros question categories.",
          "multiline": true,
          "default": "{}",
          "examples": [
            "{\"Technical Practices\": \"Speed and Agility\", \"Quality\": \"Quality\", \"Planning\": \"Alignment and Goals\"}"
          ]
        },
        "survey_name": {
          "title": "Survey Name",
          "description": "The name of the survey.",
          "type": "string"
        },
        "survey_description": {
          "title": "Survey Description",
          "description": "The description of the survey.",
          "type": "string"
        },
        "survey_type": {
          "title": "Survey Type",
          "description": "The type of survey.",
          "type": "string",
          "enum": [
            "ENPS",
            "NPS",
            "Satisfaction",
            "Custom"
          ]
        },
        "question_column_index_start": {
          "title": "Question Columns Start Index",
          "description": "The index of the first question column header.",
          "type": "integer"
        },
        "id_column_name": {
          "title": "ID Column Name",
          "description": "The name of the column that contains the record ID.",
          "type": "string"
        },
        "name_column_name": {
          "title": "Name Column Name",
          "description": "The name of the column that contains the record name.",
          "type": "string"
        },
        "email_column_name": {
          "title": "Email Column Name",
          "description": "The name of the column that contains the record email.",
          "type": "string"
        },
        "team_column_name": {
          "title": "Team Column Name",
          "description": "The name of the column that contains the record team name.",
          "type": "string"
        },
        "submitted_at_column_name": {
          "title": "Submitted At Column Name",
          "description": "The name of the column that contains the record submitted at date.",
          "type": "string",
          "format": "date-time"
        }
      }
    }
  ]
}
