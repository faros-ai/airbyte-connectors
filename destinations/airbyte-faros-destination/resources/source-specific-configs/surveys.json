{
  "title": "Surveys",
  "type": "object",
  "oneOf": [
    {
      "title": "Configuration",
      "type": "object",
      "properties": {
        "source_type": {
          "type": "string",
          "const": "Surveys",
          "order": 0
        },
        "survey_responses_table_name": {
          "title": "Survey Responses Table Name",
          "description": "Table name",
          "type": "string",
          "default": "Survey Responses",
          "order": 1
        },
        "survey_metadata_table_name": {
          "title": "Survey Metadata Table Name",
          "description": "Table name",
          "type": "string",
          "default": "Survey Metadata",
          "order": 2
        },
        "question_metadata_table_name": {
          "title": "Question Metadata Table Name",
          "description": "Table name",
          "type": "string",
          "default": "Question Metadata",
          "order": 3
        },
        "column_names_mapping": {
          "title": "Column Names Mapping",
          "description": "Indicates the columns that contain the survey data in the source records.",
          "type": "object",
          "properties": {
            "survey_name_column_name": {
              "title": "Survey Name Column Name",
              "description": "",
              "type": "string",
              "default": "Survey Name",
              "order": 0
            },
            "survey_description_column_name": {
              "title": "Survey Description Column Name",
              "description": "",
              "type": "string",
              "default": "Survey Description",
              "order": 1
            },
            "survey_type_column_name": {
              "title": "Survey Type Column Name",
              "description": "",
              "type": "string",
              "default": "Survey Type",
              "order": 2
            },
            "survey_status_column_name": {
              "title": "Survey Status Column Name",
              "description": "",
              "type": "string",
              "default": "Survey Status",
              "order": 3
            },
            "survey_started_at_column_name": {
              "title": "Survey Started At Column Name",
              "description": "",
              "type": "string",
              "default": "Survey Started At",
              "order": 4
            },
            "survey_ended_at_column_name": {
              "title": "Survey Ended At Column Name",
              "description": "",
              "type": "string",
              "default": "Survey Ended At",
              "order": 5
            },
            "respondent_name_column_name": {
              "title": "Name Column Name",
              "description": "",
              "type": "string",
              "default": "Name",
              "order": 6
            },
            "respondent_email_column_name": {
              "title": "Email Column Name",
              "description": "",
              "type": "string",
              "default": "Email",
              "order": 7
            },
            "respondent_team_id_column_name": {
              "title": "Respondent Team ID Column Name",
              "description": "",
              "type": "string",
              "default": "Team ID",
              "order": 8
            },
            "respondent_team_name_column_name": {
              "title": "Team Name Column Name",
              "description": "",
              "type": "string",
              "default": "Team Name",
              "order": 9
            },
            "question_column_name": {
              "title": "Question Column Name",
              "description": "",
              "type": "string",
              "default": "Question",
              "order": 10
            },
            "question_category_column_name": {
              "title": "Question Category Column Name",
              "description": "",
              "type": "string",
              "default": "Category",
              "order": 11
            },
            "response_type_column_name": {
              "title": "Response Type Column Name",
              "description": "",
              "type": "string",
              "default": "Response Type",
              "order": 12
            },
            "response_submitted_at_column_name": {
              "title": "Response Submitted At Column Name",
              "description": "",
              "type": "string",
              "default": "Timestamp",
              "order": 13
            }
          },
          "order": 4
        },
        "question_category_mapping": {
          "type": "string",
          "title": "Question Category Mapping",
          "description": "JSON map from source survey question categories to Faros question categories.",
          "multiline": true,
          "default": "{}",
          "examples": [
            "{\"Technical Practices\": \"SpeedAndAgility\", \"Planning\": \"AlignmentAndGoals\"}"
          ],
          "order": 5
        },
        "exclude_columns": {
          "type": "array",
          "title": "Exclude Columns",
          "description": "Columns to exclude from the source records",
          "items": {
            "type": "string"
          },
          "default": [],
          "order": 6
        }
      }
    }
  ]
}
