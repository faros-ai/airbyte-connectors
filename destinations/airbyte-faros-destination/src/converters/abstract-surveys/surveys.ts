import {createHash} from 'crypto';
import {AirbyteRecord, toDate} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';
import _ from 'lodash';

import {
  Converter,
  DestinationModel,
  DestinationRecord,
  parseObjectConfig,
  StreamContext,
} from '../converter';
import {
  QuestionCategoryMapping,
  Survey,
  SurveyCategory,
  SurveyQuestion,
  SurveyQuestionCategory,
  SurveyResponseCategory,
  SurveyStats,
  SurveyStatusCategory,
  SurveyTeam,
  SurveyUser,
} from './models';

type ColumnNameMapping = {
  survey_name_column_name?: string;
  survey_type_column_name?: string;
  survey_description_column_name?: string;
  survey_started_at_column_name?: string;
  survey_ended_at_column_name?: string;
  survey_status_column_name?: string;
  respondent_name_column_name?: string;
  respondent_email_column_name?: string;
  respondent_team_name_column_name?: string;
  respondent_team_id_column_name?: string;
  question_category_column_name?: string;
  response_type_column_name?: string;
  question_column_name?: string;
  response_submitted_at_column_name?: string;
};

export interface SurveysConfig {
  survey_responses_table_name?: string;
  survey_metadata_table_name?: string;
  question_metadata_table_name?: string;
  question_category_mapping?: QuestionCategoryMapping;
  column_names_mapping?: ColumnNameMapping;
  exclude_columns?: ReadonlyArray<string>;
}

export abstract class AbstractSurveys extends Converter {
  abstract getSurveyId(record: AirbyteRecord): string | undefined;
  abstract getTableName(record: AirbyteRecord): string | undefined;

  getSubmittedAt(record: AirbyteRecord): string | undefined {
    const submittedAtColumnName =
      this.config.column_names_mapping.response_submitted_at_column_name;
    return AbstractSurveys.getColumnValue(
      record?.record?.data?.row,
      submittedAtColumnName
    );
  }

  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'survey_Survey',
    'survey_Question',
    'survey_SurveyQuestionAssociation',
    'survey_QuestionResponse',
    'survey_User',
    'survey_Team',
    'survey_TeamMembership',
  ];

  private questionMetadata: Map<string, SurveyQuestion> = new Map();
  private surveyMetadata: Map<string, Survey> = new Map();
  private surveyStats: Map<string, SurveyStats> = new Map();

  private usersSeen = new Set<string>();
  private teamsSeen = new Set<string>();
  private surveysSeen = new Set<string>();
  private questionsSeen = new Set<string>();

  private _config: SurveysConfig = undefined;
  private _questionCategoryMapping: QuestionCategoryMapping = undefined;

  private initialize(ctx: StreamContext): void {
    this._config =
      this._config ?? ctx.config.source_specific_configs?.surveys ?? {};
    this._questionCategoryMapping =
      this._questionCategoryMapping ??
      parseObjectConfig(
        this._config?.question_category_mapping,
        'Question Category Mapping'
      ) ??
      {};
  }

  protected get config(): SurveysConfig {
    return this._config;
  }

  private get questionCategoryMapping(): QuestionCategoryMapping {
    return this._questionCategoryMapping;
  }

  async convert(
    record: AirbyteRecord,
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    this.initialize(ctx);

    const row = record?.record?.data?.row;
    const surveyId = this.getSurveyId(record);
    const tableName = this.getTableName(record);

    if (!row || !surveyId || !tableName) {
      return [];
    }

    // Be a bit more lenient with table names matching
    const normalizeTableName = (tableName: string): string => {
      return tableName.toLowerCase().split(' ').join('_');
    };

    // Question metadata
    if (
      normalizeTableName(tableName) ===
      normalizeTableName(this.config.question_metadata_table_name)
    ) {
      this.processQuestionMetadata(surveyId, row);
      return [];
    }

    // Survey metadata
    if (
      normalizeTableName(tableName) ===
      normalizeTableName(this.config.survey_metadata_table_name)
    ) {
      this.processSurveyMetadata(surveyId, row);
      return [];
    }

    // Survey response
    if (
      normalizeTableName(tableName) ===
      normalizeTableName(this.config.survey_responses_table_name)
    ) {
      const responseId = this.id(record);
      const submittedAt = this.getSubmittedAt(record);

      const questions = this.getFilteredQuestions(row);
      const res = this.processResponse(
        surveyId,
        row,
        responseId,
        submittedAt,
        questions
      );

      // Update survey stats for pushing on processing complete
      this.updateSurveyStats(surveyId, questions);

      return res;
    }

    return [];
  }

  private processQuestionMetadata(surveyId: string, row: any): void {
    const questionWithMetadata = this.getQuestionWithMetadata(surveyId, row);

    if (!questionWithMetadata) {
      return;
    }

    this.questionMetadata.set(questionWithMetadata.uid, questionWithMetadata);
  }

  private processSurveyMetadata(surveyId: string, row: any) {
    const surveyRecord = this.getSurveyRecord(surveyId, row);
    this.surveyMetadata.set(surveyRecord.uid, surveyRecord);
  }

  private processResponse(
    surveyId: string,
    row: any,
    responseId: string,
    submittedAt: string,
    questions: string[]
  ) {
    const res = [];

    const surveyUser: SurveyUser | undefined = this.getSurveyUser(row);
    const surveyTeam: SurveyTeam | undefined = this.getSurveyTeam(row);

    if (surveyUser && !this.usersSeen.has(surveyUser.uid)) {
      this.usersSeen.add(surveyUser.uid);
      res.push({
        model: 'survey_User',
        record: surveyUser,
      });
    }

    if (surveyTeam && !this.teamsSeen.has(surveyTeam.uid)) {
      this.teamsSeen.add(surveyTeam.uid);
      res.push({
        model: 'survey_Team',
        record: surveyTeam,
      });
    }

    if (surveyUser && surveyTeam) {
      res.push({
        model: 'survey_TeamMembership',
        record: {
          member: _.pick(surveyUser, ['uid', 'source']),
          team: _.pick(surveyTeam, ['uid', 'source']),
        },
      });
    }

    res.push(
      ...this.getResponseRecords(
        surveyId,
        row,
        responseId,
        submittedAt,
        questions,
        surveyUser,
        surveyTeam
      )
    );
    return res;
  }

  /** Upsert surveys to add stats and survey questions records to include metadata (question category and response type) **/
  async onProcessingComplete(
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const updateRecords = [];
    // Add survey questions upsert mutations
    this.questionMetadata.forEach((surveyQuestion) => {
      const questionRecord = {
        model: 'survey_Question__Update',
        record: {
          at: Date.now(),
          where: {uid: surveyQuestion.uid, source: surveyQuestion.source},
          mask: ['questionCategory', 'responseType'],
          patch: {
            questionCategory: surveyQuestion.questionCategory,
            responseType: surveyQuestion.responseType,
          },
        },
      };
      updateRecords.push(questionRecord);
    });

    this.surveyStats.forEach((surveyStats, surveyId) => {
      const surveyRecord = {
        model: 'survey_Survey__Update',
        record: {
          at: Date.now(),
          where: {uid: surveyId, source: this.streamName.source},
          mask: ['stats'],
          patch: {
            stats: surveyStats,
          },
        },
      };
      updateRecords.push(surveyRecord);
    });

    this.surveyMetadata.forEach((survey) => {
      updateRecords.push({
        model: 'survey_Survey__Update',
        record: {
          at: Date.now(),
          where: {
            uid: survey.uid,
            source: this.streamName.source,
          },
          mask: [
            'name',
            'description',
            'type',
            'status',
            'startedAt',
            'endedAt',
          ],
          patch: {
            name: survey.name,
            description: survey.description,
            type: survey.type,
            status: survey.status,
            startedAt: survey.startedAt,
            endedAt: survey.endedAt,
          },
        },
      });
    });

    return updateRecords;
  }

  private getQuestionWithMetadata(surveyId: string, row: any): SurveyQuestion {
    const question = AbstractSurveys.getColumnValue(
      row,
      this.config.column_names_mapping.question_column_name
    );

    if (!question) {
      return undefined;
    }

    const category = AbstractSurveys.getColumnValue(
      row,
      this.config.column_names_mapping.question_category_column_name
    );
    const responseType = AbstractSurveys.getColumnValue(
      row,
      this.config.column_names_mapping.response_type_column_name
    );

    return {
      uid: AbstractSurveys.createQuestionUid(surveyId, question),
      source: this.source,
      questionCategory: category
        ? Utils.toCategoryDetail(
            SurveyQuestionCategory,
            category,
            this.questionCategoryMapping
          )
        : null,
      responseType: responseType
        ? Utils.toCategoryDetail(SurveyResponseCategory, responseType)
        : null,
    };
  }

  private getResponseRecords(
    surveyId: string,
    row: any,
    responseId: string,
    submittedAt: string,
    questions: string[],
    surveyUser?: SurveyUser,
    surveyTeam?: SurveyTeam
  ) {
    return questions.flatMap((question, index) => {
      const surveyRecord = this.getSurveyRecord(surveyId, row);
      const questionId = AbstractSurveys.createQuestionUid(surveyId, question);

      const questionRecord = {
        uid: questionId,
        source: this.source,
        question,
      };

      const surveyQuestionAssociationRecord = {
        survey: {uid: surveyId, source: this.source},
        question: {uid: questionRecord.uid, source: this.source},
        order: index + 1,
      };

      const questionResponse = {
        model: 'survey_QuestionResponse',
        record: {
          uid: responseId,
          source: this.source,
          submittedAt,
          response: AbstractSurveys.getColumnValue(row, question).toString(),
          surveyQuestion: {
            survey: {uid: surveyId, source: this.source},
            question: {uid: questionRecord.uid, source: this.source},
          },
          respondent: surveyUser
            ? {uid: surveyUser.uid, source: this.source}
            : null,
          team: surveyTeam ? {uid: surveyTeam.uid, source: this.source} : null,
        },
      };

      const res: DestinationRecord[] = [questionResponse];

      if (!this.questionsSeen.has(questionRecord.uid)) {
        this.questionsSeen.add(questionRecord.uid);
        res.push(
          ...[
            {
              model: 'survey_Question',
              record: questionRecord,
            },
            {
              model: 'survey_SurveyQuestionAssociation',
              record: surveyQuestionAssociationRecord,
            },
          ]
        );
      }

      if (!this.surveysSeen.has(surveyId)) {
        this.surveysSeen.add(surveyId);
        res.push({
          model: 'survey_Survey',
          record: surveyRecord,
        });
      }

      return res;
    });
  }

  private getSurveyRecord(surveyId: string, row: any): Survey {
    const surveyData = this.getSurveyData(row);
    return {
      uid: surveyId,
      source: this.source,
      status: surveyData.status
        ? Utils.toCategoryDetail(SurveyStatusCategory, surveyData.status)
        : null,
      type: surveyData.type
        ? Utils.toCategoryDetail(SurveyCategory, surveyData.type)
        : null,
      name: surveyData.name,
      description: surveyData.description,
      startedAt: surveyData.startedAt ? toDate(surveyData.startedAt) : null,
      endedAt: surveyData.endedAt ? toDate(surveyData.endedAt) : null,
    };
  }

  private updateSurveyStats(surveyId: string, questions: string[]) {
    const stats = this.surveyStats.get(surveyId) || {
      questionCount: questions.length,
      invitationCount: 0, // Invitation count is left at 0 as we don't have this data from Airtable
      responseCount: 0,
    };
    stats.responseCount += 1;
    this.surveyStats.set(surveyId, stats);
  }

  private getFilteredQuestions(row: any): string[] {
    const excludeColumns = this.config.exclude_columns ?? [];
    return Object.keys(row).filter(
      (question) =>
        ![
          this.config.column_names_mapping.survey_name_column_name,
          this.config.column_names_mapping.survey_type_column_name,
          this.config.column_names_mapping.survey_description_column_name,
          this.config.column_names_mapping.survey_started_at_column_name,
          this.config.column_names_mapping.survey_ended_at_column_name,
          this.config.column_names_mapping.survey_status_column_name,
          this.config.column_names_mapping.respondent_name_column_name,
          this.config.column_names_mapping.respondent_email_column_name,
          this.config.column_names_mapping.respondent_team_name_column_name,
          this.config.column_names_mapping.respondent_team_id_column_name,
          this.config.column_names_mapping.response_submitted_at_column_name,
          ...excludeColumns,
        ].includes(question)
    );
  }

  private getSurveyData(row: any): {
    name: string | null;
    type: string | null;
    description: string | null;
    startedAt: string | null;
    endedAt: string | null;
    status: string | null;
  } {
    return {
      name:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.survey_name_column_name
        ) ?? null,
      type:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.survey_type_column_name
        ) ?? null,
      description:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.survey_description_column_name
        ) ?? null,
      startedAt:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.survey_started_at_column_name
        ) ?? null,
      endedAt:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.survey_ended_at_column_name
        ) ?? null,
      status:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.survey_status_column_name
        ) ?? null,
    };
  }

  private getSurveyUser(row: any): SurveyUser | undefined {
    const uid = AbstractSurveys.getColumnValue(
      row,
      this.config.column_names_mapping.respondent_email_column_name
    );

    if (!uid) {
      return undefined;
    }

    return {
      uid,
      source: this.source,
      email: uid,
      name:
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.respondent_name_column_name
        ) ?? null,
    };
  }

  private getSurveyTeam(row: any): SurveyTeam | undefined {
    const uid =
      AbstractSurveys.getColumnValue(
        row,
        this.config.column_names_mapping.respondent_team_id_column_name
      ) ??
      AbstractSurveys.getTeamUid(
        AbstractSurveys.getColumnValue(
          row,
          this.config.column_names_mapping.respondent_team_name_column_name
        )
      );

    if (!uid) {
      return undefined;
    }

    return {
      uid,
      source: this.source,
      name: AbstractSurveys.getColumnValue(
        row,
        this.config.column_names_mapping.respondent_team_name_column_name
      ),
    };
  }

  static createQuestionUid(surveyId: string, question: string) {
    return AbstractSurveys.digest(`${surveyId}-${question}`);
  }

  static getTeamUid(teamName: string): string | undefined {
    if (!teamName) {
      return undefined;
    }

    return AbstractSurveys.digest(teamName);
  }

  static digest(input: string): string {
    return createHash('sha256').update(input).digest('hex');
  }

  // Get column value either from string or an array of single string (lookup result)
  static getColumnValue(row: any, columnName: string): string | null {
    const columnValue = row[columnName];
    if (!columnValue) {
      return null;
    }
    if (Array.isArray(columnValue)) {
      return columnValue.length === 0 ? null : columnValue[0];
    }
    return columnValue;
  }
}
