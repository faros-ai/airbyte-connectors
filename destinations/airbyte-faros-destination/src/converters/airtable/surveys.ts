import {AirbyteRecord} from 'faros-airbyte-cdk';

import {
  AbstractSurveys,
} from '../abstract-surveys/surveys';

export class Surveys extends AbstractSurveys {
  source = 'Airtable';

  id(record: AirbyteRecord): any {
    return record?.record?.data?._airtable_id;
  }

  static getBaseId(fqTableId: string): string {
    // Get the first part of the fully qualified table id, which corresponds to the Airtable base id
    // E.g., appwVNmuUAPCIxzSZ/tblWFFSCLxi0gVtkU -> appwVNmuUAPCIxzSZ
    return fqTableId.split('/')[0];
  }

  static getTableName(fqTableName: string): string {
    // Get the second part of the fully qualified table name, which corresponds to the table name
    // E.g., my_surveys/Survey Responses => Survey Responses
    return fqTableName.split('/')[1];
  }

  getSurveyId(record: AirbyteRecord): string | undefined {
    const fqTableId: string = record?.record?.data?._airtable_table_id;
    if (!fqTableId) {
      return undefined;
    }

    return Surveys.getBaseId(fqTableId);
  }

  getTableName(record: AirbyteRecord): string | undefined {
    const fqTableName: string = record?.record?.data?._airtable_table_name;
    if (!fqTableName) {
      return undefined;
    }

    return Surveys.getTableName(fqTableName);
  }

  getSubmittedAt(record: AirbyteRecord): string | undefined {
    return super.getSubmittedAt(record) ?? record?.record?.data?._airtable_created_time;
  }
}
