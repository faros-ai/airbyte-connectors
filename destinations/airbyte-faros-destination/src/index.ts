import {Command} from 'commander';

import {FarosDestinationRunner} from './destination-runner';

/**
 * The main entry point
 *
 * You can optionally provide options `exitOverride` and `configureOutput` to
 * override default behavior of the CLI when it detects an error and where it
 * writes to respectively. See https://github.com/tj/commander.js#override-exit-and-output-handling
 * for detailed info on how these options work.
 */
export function mainCommand(options?: {
  exitOverride?: boolean;
  suppressOutput?: boolean;
}): Command {
  const destinationRunner = new FarosDestinationRunner();
  const program = destinationRunner.mainCommand();

  if (options?.exitOverride) {
    program.exitOverride();
  }
  if (options?.suppressOutput) {
    program.configureOutput({
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      writeOut: () => {},
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      writeErr: () => {},
    });
  }

  return program;
}

export {Edition, InvalidRecordStrategy} from './common/types';
export {
  Converter,
  DestinationRecord,
  DestinationModel,
  StreamContext,
  StreamName,
  StreamNameSeparator,
} from './converters/converter';
export {FarosDestinationRunner} from './destination-runner';
