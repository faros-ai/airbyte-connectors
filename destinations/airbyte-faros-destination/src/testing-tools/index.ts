export * from './cli';
export * from './testing-tools';
export * from './utils';
