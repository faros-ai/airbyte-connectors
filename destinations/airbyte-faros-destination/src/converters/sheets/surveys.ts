import {AirbyteRecord} from 'faros-airbyte-cdk';

import {AbstractSurveys} from '../abstract-surveys/surveys';

export class Surveys extends AbstractSurveys {
  source = 'Sheets';

  id(record: AirbyteRecord): any {
    return record?.record?.data?.id;
  }

  getSurveyId(record: AirbyteRecord): string | undefined {
    return record.record.data.sheetId;
  }

  getTableName(record: AirbyteRecord): string | undefined {
    return record.record.data.sheetName;
  }
}
