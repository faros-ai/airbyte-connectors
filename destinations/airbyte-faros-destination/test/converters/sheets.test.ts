import {AirbyteLogger, AirbyteRecord} from 'faros-airbyte-cdk';
import {getLocal} from 'mockttp';

import {StreamContext} from '../../src';
import {SurveysConfig} from '../../src/converters/abstract-surveys/surveys';
import {Surveys} from '../../src/converters/sheets/surveys';
import {initMockttp} from '../testing-tools';

describe('sheets', () => {
  const mockttp = getLocal({debug: false, recordTraffic: false});
  let converter: Surveys;

  beforeEach(async () => {
    await initMockttp(mockttp);
    jest.spyOn(Date, 'now').mockImplementation(() => 1697245567000);
    converter = new Surveys();
  });

  afterEach(async () => {
    await mockttp.stop();
    jest.restoreAllMocks();
  });

  describe('survey responses', () => {
    const surveysConfig: SurveysConfig = {
      survey_responses_table_name: 'Survey Responses',
      survey_metadata_table_name: 'Survey Metadata',
      question_metadata_table_name: 'Question Metadata',
      question_category_mapping: {},
      column_names_mapping: {
        survey_name_column_name: 'Survey Name',
        survey_type_column_name: 'Survey Type',
        survey_started_at_column_name: 'Survey Started At',
        survey_ended_at_column_name: 'Survey Ended At',
        survey_status_column_name: 'Survey Status',
        survey_description_column_name: 'Survey Description',
        respondent_name_column_name: 'Name',
        respondent_email_column_name: 'Email',
        respondent_team_name_column_name: 'Team Name',
        respondent_team_id_column_name: 'Team ID',
        question_category_column_name: 'Category',
        response_type_column_name: 'Response Type',
        question_column_name: 'Question',
        response_submitted_at_column_name: 'Timestamp',
      },
    };

    test('basic response convert and getters', async () => {
      const record = AirbyteRecord.make('surveys', {
        id: '10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4_survey_responses_1',
        sheetId: '10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4',
        sheetName: 'Survey Responses',
        row: {
          'How much do you like ice cream?': 5,
          'Team Name': 'X',
          Timestamp: '2023-10-09T14:09:37.000Z',
        },
      });
      const ctx = new StreamContext(
        new AirbyteLogger(),
        {
          edition_configs: {},
          source_specific_configs: {
            surveys: surveysConfig,
          },
        },
        {}
      );
      const res = await converter.convert(record, ctx);
      expect(res).toMatchSnapshot();

      const id = converter.id(record);
      expect(id).toEqual(
        '10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4_survey_responses_1'
      );

      const submittedAt = converter.getSubmittedAt(record);
      expect(submittedAt).toEqual('2023-10-09T14:09:37.000Z');

      const surveyId = converter.getSurveyId(record);
      expect(surveyId).toEqual('10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4');

      const tableName = converter.getTableName(record);
      expect(tableName).toEqual('Survey Responses');
    });
  });
});
