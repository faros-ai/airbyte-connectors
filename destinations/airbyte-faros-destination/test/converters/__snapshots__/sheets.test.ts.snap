// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sheets survey responses basic response convert and getters 1`] = `
Array [
  Object {
    "model": "survey_Team",
    "record": Object {
      "name": "X",
      "source": "Sheets",
      "uid": "4b68ab3847feda7d6c62c1fbcbeebfa35eab7351ed5e78f4ddadea5df64b8015",
    },
  },
  Object {
    "model": "survey_QuestionResponse",
    "record": Object {
      "respondent": null,
      "response": "5",
      "source": "Sheets",
      "submittedAt": "2023-10-09T14:09:37.000Z",
      "surveyQuestion": Object {
        "question": Object {
          "source": "Sheets",
          "uid": "3dbd78e458bab6402e2c370d65c3da160e8ff3f68585e176e0c0a0dd4801c56d",
        },
        "survey": Object {
          "source": "Sheets",
          "uid": "10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4",
        },
      },
      "team": Object {
        "source": "Sheets",
        "uid": "4b68ab3847feda7d6c62c1fbcbeebfa35eab7351ed5e78f4ddadea5df64b8015",
      },
      "uid": "10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4_survey_responses_1",
    },
  },
  Object {
    "model": "survey_Question",
    "record": Object {
      "question": "How much do you like ice cream?",
      "source": "Sheets",
      "uid": "3dbd78e458bab6402e2c370d65c3da160e8ff3f68585e176e0c0a0dd4801c56d",
    },
  },
  Object {
    "model": "survey_SurveyQuestionAssociation",
    "record": Object {
      "order": 1,
      "question": Object {
        "source": "Sheets",
        "uid": "3dbd78e458bab6402e2c370d65c3da160e8ff3f68585e176e0c0a0dd4801c56d",
      },
      "survey": Object {
        "source": "Sheets",
        "uid": "10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4",
      },
    },
  },
  Object {
    "model": "survey_Survey",
    "record": Object {
      "description": null,
      "endedAt": null,
      "name": null,
      "source": "Sheets",
      "startedAt": null,
      "status": null,
      "type": null,
      "uid": "10hWFrCr5c0J7wUjoPODgd7jt6pRjPBGJTtD6y_z2He4",
    },
  },
]
`;
